plugins/lttng-utils/debug-info.c

```diff
@@ -726,22 +913,933 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
 		 * of the dlopen family are called (e.g. dlmopen) and when
 		 * library are transitively loaded.
 		 */
-		handle_lib_load_event(err, debug_info, event);
+		handle_event_lib_load(debug_info, event);
 	} else if (q_event_name == debug_info->q_statedump_start) {
 		/* Start state dump */
-		handle_statedump_start(err, debug_info, event);
+		handle_event_statedump_start(debug_info, event);
 	} else if (q_event_name == debug_info->q_statedump_debug_link) {
 		/* Debug link info */
-		handle_statedump_debug_link_event(err, debug_info, event);
+		handle_event_statedump_debug_link(debug_info, event);
 	} else if (q_event_name == debug_info->q_statedump_build_id) {
 		/* Build ID info */
-		handle_statedump_build_id_event(err, debug_info, event);
+		handle_event_statedump_build_id(debug_info, event);
 	} else if (q_event_name == debug_info-> q_lib_unload) {
-		handle_lib_unload_event(err, debug_info, event);
+		handle_event_lib_unload(debug_info, event);
 	}
 
-end_put_class:
-	bt_event_class_put_ref(event_class);
-end:
 	return;
 }
+
+static
+void destroy_debug_info_comp(struct debug_info_component *debug_info)
+{
+	free(debug_info->arg_debug_info_field_name);
+	g_free(debug_info);
+}
+
+static inline
+void fill_debug_info_bin_field(struct debug_info_source *dbg_info_src,
+		bool full_path, GString *curr_field_str)
+{
+	if (full_path) {
+		g_string_printf(curr_field_str, "%s%s",
+				dbg_info_src->bin_path, dbg_info_src->bin_loc);
+	} else {
+		g_string_printf(curr_field_str, "%s%s",
+				dbg_info_src->short_bin_path,
+				dbg_info_src->bin_loc);
+	}
+}
+
+static inline
+void fill_debug_info_func_field(struct debug_info_source *dbg_info_src,
+		GString *curr_field_str)
+{
+	if (dbg_info_src->func) {
+		g_string_printf(curr_field_str, "%s", dbg_info_src->func);
+	}
+}
+
+static inline
+void fill_debug_info_src_field(struct debug_info_source *dbg_info_src,
+		bool full_path, GString *curr_field_str)
+{
+	if (dbg_info_src->src_path) {
+		if (full_path) {
+			g_string_printf(curr_field_str, "%s:%" PRId64,
+					dbg_info_src->src_path,
+					dbg_info_src->line_no);
+		} else {
+			g_string_printf(curr_field_str, "%s:%" PRId64,
+					dbg_info_src->short_src_path,
+					dbg_info_src->line_no);
+		}
+	}
+}
+
+static
+void fill_debug_info_field(struct debug_info *debug_info, int64_t vpid,
+		uint64_t ip, bt_field *debug_info_field)
+{
+	struct debug_info_source * dbg_info_src;
+	const bt_field_class *debug_info_fc;
+	bt_field_status status;
+	uint64_t i, nb_member;
+	GString *curr_field_str;
+
+	BT_ASSERT(bt_field_get_class_type(debug_info_field) ==
+			BT_FIELD_CLASS_TYPE_STRUCTURE);
+
+	debug_info_fc = bt_field_borrow_class_const(debug_info_field);
+
+	nb_member = bt_field_class_structure_get_member_count(debug_info_fc);
+	BT_ASSERT(nb_member == 3);
+
+	dbg_info_src = debug_info_query(debug_info, vpid, ip);
+
+	/* Start with an empty string. */
+	curr_field_str = g_string_new(NULL);
```

Comment:
In all cases, we need a string, empty or not. If it needs to be empty, because debug_info_query() returned NULL, we could use a static char array ("\0") to avoid the allocation. Also the loop right after that could be rewrote into a series of call to ` bt_field_class_structure_borrow_member_by_name_const()`. I will look into all this.

Reply:
deI don't understand how `bt_field_class_structure_borrow_member_by_name_const()` is related here.

But getting rid of the `GString` and using `bt_field_string_append()` in the `fill_debug_info_*()` functions would solve this. Just don't forget to call `bt_field_string_clear()` initially in `fill_debug_info_field()`.

Reply:
My `bt_field_class_structure_borrow_member_by_name_const()` comment was unrelated to your initial comment, it was just something I realized when looking at your comment.

Reply:
I made it so that there are no allocation in that loop on the fast path. I will make sure you have a look during the next review.